src/dsmr/parser.h

```diff
@@ -316,27 +316,43 @@ struct P1Parser {
 
     // Look for ! that terminates the data
     const char *data_end = data_start;
-    uint16_t crc = _crc16_update(0, *str); // Include the / in CRC
-    while (data_end < str + n && *data_end != '!') {
-      crc = _crc16_update(crc, *data_end);
-      ++data_end;
-    }
+    const char *next = NULL;
+    if(checksum) {
+      uint16_t crc = _crc16_update(0, *str); // Include the / in CRC
+      while (data_end < str + n && *data_end != '!') {
+        crc = _crc16_update(crc, *data_end);
+        ++data_end;
+      }
 
-    if (data_end >= str + n)
-      return res.fail(F("No checksum found"), data_end);
+      if (data_end >= str + n)
+        return res.fail(F("No checksum found"), data_end);
 
-    crc = _crc16_update(crc, *data_end); // Include the ! in CRC
+      crc = _crc16_update(crc, *data_end); // Include the ! in CRC
 
-    ParseResult<uint16_t> check_res = CrcParser::parse(data_end + 1, str + n);
-    if (check_res.err)
-      return check_res;
+      ParseResult<uint16_t> check_res = CrcParser::parse(data_end + 1, str + n);
+      if (check_res.err)
+        return check_res;
 
-    // Check CRC
-    if (check_res.result != crc)
-      return res.fail(F("Checksum mismatch"), data_end + 1);
+      // Check CRC
+      if (check_res.result != crc)
+        return res.fail(F("Checksum mismatch"), data_end + 1);
+
+      next = check_res.next;
+    } else {
+      // There's no CRC check, still need to know where the message ends (!)
+      while (data_end < str + n && *data_end != '!') {
+        ++data_end;
+      }
+
+      // Skip to end-of-line, everything afther that is not yet processed.
+      next = data_end;
+      while (next < str + n && *next != '\r' && *next != '\n') {
+        ++next;
+      }
```

Comment:
Is this needed? Is there a CRLF after the !? If I read the v4 spec correctly, there should be no CRLF after the last line (which I think is the checksum line in v4, and the ! line in v3)? Are you seeing a CRLF in your messages?

Reply:
Hmm, you could be right, I've checked the v3 spec and there's no CRLF. Need to check if that's somehow added to my recording.